Default to local_freetype builds.

anntzer opened the pull request:
This would make builds from source work straightforwardly (including
for testing purposes) and simplify the build instructions for new
contributors -- see changes in build instructions.

~~Note that there is intentionally no corresponding entry in setup.cfg
because it is unclear whether e.g. a set but *empty* MPLSYSTEMFREETYPE
environment variable should override a hypothetical system_freetype =
True in setup.cfg, and because this would make the error message in
checkdep_freetype.c more complex (do we tell the use to unset the
environment variable?  or should setting it to "0" or "false" be the
same as unsetting it?).  The design here keeps things simple by having a
single "on" switch.~~

Note that there is intentionally no corresponding environment variable
because it is unclear whether e.g. a set but *empty* MPLSYSTEMFREETYPE
environment variable should override a system_freetype = True
in setup.cfg, and because this would make the error message in
checkdep_freetype.c more complex (do we tell the use to unset the
environment variable? or should setting it to "0" or "false" be the same
as unsetting it?).  The design here keeps things simple by having a
single "on" switch.

The config entry is under [libs] to allow for a future system_qhull
entry.

As suggested in https://github.com/matplotlib/matplotlib/pull/11984#issuecomment-417572465 we could have a similar switch for libqhull.

attn @QuLogic @sandrotosi @felixonmars @ArchangeGabriel (the linux packagers).

## PR Summary

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

anntzer commented:
I have a mild preference for the environment variable (there's less friction setting/unsetting it from the command line than opening a config file, editing it, and saving) but can switch to the cfg approach if the consensus is on that side.  Any reason for preferring it?

felixonmars commented:
It looks to me that all the other environment variables are mostly used for runtime, while setup.cfg switches are mostly for the build script. The first comment in setup.py also states "The matplotlib build options can be modified with a setup.cfg file" and does not mention environment variables. It would be more consistent if the freetype build option keeps that way.

anntzer commented:
Fair enough.  Let's give some time for others to chime in (the implementation is not the problem here :)).

sandrotosi commented:
I'm probably +0 on using setup.cfg (mostly because in debian we already do :) )

what i'm mildly concerned about is what's gonna happen with the baseline images and unittests : did you look into the impact that may have and how to address it?

anntzer commented:
The situation is unchanged for baseline images: you still need to have a ft2.6.1 build for baseline images to match exactly; as Debian uses (I guess) a more recent ft you'll need to keep whatever patch you are using to make tests pass (likely, raising the tolerance of image comparison tests).

ArchangeGabriel commented:
Regarding that last point, as said in https://github.com/matplotlib/matplotlib/pull/14170#issuecomment-491580955 and following comments, skipping the baseline image tests could also be a possibility (I’m not sure relaxed tests are way better than no tests).

And obviously I agree with @felixonmars that keeping everything in setup.cfg makes more sense.

anntzer commented:
OK, I switched to a libs.system_freetype entry in setup.cfg.  (I moved it out of `[tests]` to prepare for a similar system_qhull entry, and they cannot be under e.g. `[build]` because they would be interpreted as parameters to `python setup.py build` (these should really have been in a separate file but heh).)

tacaswell commented:
We should also get feedback from @cgohlke @msarahan and track down someone from brew.

Given that this will make it much easier for new contributors to get a working build I am :+1:.

It does put a bit more burden on our downstream packagers, but it is likely they are maintaining patches / some scripts around us anyway so I don't expect it to be be a huge burden (but they will need to know it's coming!) so that is a -0 for me.

anntzer commented:
You're :+1: but -0? :)

msarahan commented:
Seems fine to me.  We can set env vars or patch setup.cfg appropriately.  No preference.  Thanks for the heads-up @tacaswell

cgohlke commented:
That's ok with me.

tacaswell commented:
@anntzer sorry for not being clear. On one concern I am :+1: on the other I am -0 which is still net :+1: .
